xline/src/storage/db.rs

```diff
@@ -1,69 +1,124 @@
+use super::{revision::Revision, storage_abstraction::Storage};
 use crate::rpc::KeyValue;
 
 /// Database to store revision to kv mapping
 #[derive(Debug)]
-pub(crate) struct DB {
+pub(crate) struct DB<S>
+where
+    S: Storage,
+{
     /// internal storage of `DB`
-    storage: Mutex<HashMap<Revision, KeyValue>>,
+    storage: S,
 }
 
-impl DB {
+impl<S> DB<S>
+where
+    S: Storage,
+{
     /// New `DB`
-    pub(crate) fn new() -> Self {
-        Self {
-            storage: Mutex::new(HashMap::new()),
-        }
+    pub(crate) fn new(storage: S) -> Self {
+        Self { storage }
     }
 
     /// Insert a `KeyValue`
-    pub(crate) fn insert(&self, revision: Revision, kv: KeyValue) -> Option<KeyValue> {
-        self.storage.lock().insert(revision, kv)
+    pub(crate) fn insert(&self, revision: Revision, kv: &KeyValue) -> Result<(), ExecuteError> {
+        let key = revision.encode_to_vec();
+        let value = kv.encode_to_vec();
+        let _prev_val = self.storage.insert(key, value).map_err(|e| {
+            ExecuteError::InvalidCommand(format!("Failed to insert revision {:?}: {}", revision, e))
+        })?;
+        Ok(())
     }
 
-    /// Get a list of `KeyValue`
-    pub(crate) fn get_values(&self, revisions: &[Revision]) -> Vec<KeyValue> {
-        let storage = self.storage.lock();
-        revisions
+    /// Get a list of `KeyValue` by `Revision`
+    pub(crate) fn get_values(&self, revisions: &[Revision]) -> Result<Vec<KeyValue>, ExecuteError> {
+        let keys = revisions
             .iter()
-            .filter_map(|revision| storage.get(revision).cloned())
-            .collect()
+            .map(Revision::encode_to_vec)
+            .collect::<Vec<_>>();
+        let values = self
+            .storage
+            .batch_get(&keys)
+            .map_err(|e| {
+                ExecuteError::InvalidCommand(format!(
+                    "Failed to get revisions {:?}: {}",
+                    revisions, e
+                ))
+            })?
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>();
+        assert_eq!(values.len(), revisions.len(), "Index doesn't match with DB");
+        let kvs = values
+            .into_iter()
+            .map(|v| KeyValue::decode(v.as_slice()))
+            .collect::<Result<_, _>>()
+            .map_err(|e| {
+                ExecuteError::InvalidCommand(format!(
+                    "Failed to decode revisions {:?}: {}",
+                    revisions, e
+                ))
+            })?;
+        Ok(kvs)
     }
 
     /// Mark deletion for keys
-    /// TODO support don't return `prev_kvs`
-    pub(crate) fn mark_deletions(&self, revisions: &[(Revision, Revision)]) -> Vec<KeyValue> {
-        let mut storage = self.storage.lock();
+    pub(crate) fn mark_deletions(
+        &self,
+        revisions: &[(Revision, Revision)],
+    ) -> Result<Vec<KeyValue>, ExecuteError> {
         let prev_kvs: Vec<KeyValue> = revisions
```

Comment:
We grab the storage lock only once before and now twice. Will there be concurrent issues?

Reply:
Yes, moved the lock out of memory, then we can lock it only once like before